core/src/main/java/org/apache/iceberg/MetadataTableUtils.java

```diff
@@ -109,4 +117,68 @@ public static Table createMetadataTableInstance(
   private static String metadataTableName(String tableName, MetadataTableType type) {
     return tableName + (tableName.contains("/") ? "#" : ".") + type.name().toLowerCase(Locale.ROOT);
   }
+
+  static class StructWithDerivedColumns implements StructLike {
```

Comment:
Initially I think this made sense because we had only one column.  But with two, it becomes un-managable.  Can we investigate whether StructProjection will work here?  Then we can remove both this and the MetricsUtil struct class.

Reply:
Discussed with @szehon-ho offline and adopted index based projection map similar to StructProjection. However I realized there's [some logic in BaseFile](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/BaseFile.java#L106-L120) are really similar to what we want. However there's some thing I dont quite get for partitionData projection when when one of the partition column get dropped as part of partition evolution. So far I am leave the change as is for BaseFiles. If everything looks right to you, I will adopt the computed columns change for BaseEntries files as well.